include/commissioner/network_diag_data.hpp

```diff
@@ -0,0 +1,187 @@
+ */
+
+#ifndef OT_COMM_NETWORK_DIAG_TLVS_HPP_
+#define OT_COMM_NETWORK_DIAG_TLVS_HPP_
+
+#include <cstddef>
+#include <cstdint>
+#include <stdbool.h>
+#include <string>
+#include <vector>
+
+#include "defines.hpp"
+#include "error.hpp"
+
+#define CHILD_TABLE_ENTRY_BYTES 3
+#define IPV6_ADDRESS_BYTES 16
+#define LEADER_DATA_BYTES 8
+#define MAC_COUNTERS_BYTES 36
+#define RLOC16_BYTES 2
+#define ROUTER_ID_MASK_BYTES 8
+
+namespace ot {
+
+namespace commissioner {
+
+/**
+ * @brief Mode Data
+ */
+struct ModeData
+{
+    bool mIsRxOnWhenIdleMode          = false;
+    bool mIsMtd                       = false;
+    bool mIsStableNetworkDataRequired = false;
+};
+
+/**
+ * @brief Child Entry in Child Table
+ */
+struct ChildTableEntry
+{
+    uint32_t mTimeout             = 0;
+    uint8_t  mIncomingLinkQuality = 0;
+    uint16_t mChildId             = 0;
+    ModeData mModeData;
+};
+
+/**
+ * @brief IPv6 Addresses List
+ */
+struct Ipv6AddressList
+{
+    std::vector<std::string> mIpv6Addresses;
+};
+
+/**
+ * @brief Leader Data
+ */
+struct LeaderData
+{
+    uint32_t mPartitionId       = 0;
+    uint8_t  mWeighting         = 0;
+    uint8_t  mDataVersion       = 0;
+    uint8_t  mStableDataVersion = 0;
+    uint8_t  mRouterId          = 0;
+};
+
+/**
+ * @brief Route Data Entry of RouteData in Route64
+ */
+struct RouteDataEntry
+{
+    uint8_t mRouterId            = 0;
+    uint8_t mOutgoingLinkQuality = 0;
+    uint8_t mIncomingLinkQuality = 0;
+    uint8_t mRouteCost           = 0;
+};
+
+/**
+ * @brief Route64 Data
+ */
+struct Route64
+{
+    uint8_t                     mIdSequence = 0;
+    ByteArray                   mMask;
+    std::vector<RouteDataEntry> mRouteData;
+};
+
+/**
+ * @brief Child IPv6 Address List
+ */
+struct ChildIpv6AddressList
+{
+    uint16_t        mRloc16 = 0;
+    Ipv6AddressList mIpv6AddressList;
+};
+
+/**
+ * @brief MAC Counters
+ */
+struct MacCounters
+{
+    uint32_t mIfInUnknownProtos  = 0;
+    uint32_t mIfInErrors         = 0;
+    uint32_t mIfOutErrors        = 0;
+    uint32_t mIfInUcastPkts      = 0;
+    uint32_t mIfInBroadcastPkts  = 0;
+    uint32_t mIfInDiscards       = 0;
+    uint32_t mIfOutUcastPkts     = 0;
+    uint32_t mIfOutBroadcastPkts = 0;
+    uint32_t mIfOutDiscards      = 0;
+};
+
+/**
+ * @brief network diagnostic data in TMF
+ *
+ * Each data field of Diagnostic TLVs is optional. The field is
+ * meaningful only when associative PresentFlags is included in
+ * `mPresentFlags`.
+ */
+struct NetDiagData
+{
+    ByteArray                    mExtMacAddress;
+    uint16_t                     mMacAddress = 0;
+    ModeData                     mMode;
+    Route64                      mRoute64;
+    LeaderData                   mLeaderData;
+    Ipv6AddressList              mIpv6AddressList;
+    std::vector<ChildTableEntry> mChildTable;
+    ByteArray                    mEui64;
+    ByteArray                    mTlvTypeList;
+    ChildIpv6AddressList         mChildIpv6AddressList;
```

Comment:
```suggestion
    int mChildRloc16;
    std::vector<std::string>         mChildAddrs;
```

And remove the `ChildIpv6AddressList` class

Reply:
no, we cannot separate the `ChildIpv6AddressList` to `mChildRloc16` and `mChildAddrs` here. Because a Thread device may have more than one child, your change cannot support it.

Reply:
I see why you want the `ChildIpv6AddressList` now, you want to use a single present flag for a single member field, right? I think it's okay for this purpose. But still you don't need the `Ipv6AddressList` class, pleae delete that one.

Besides, I think the name "ChildAddressInfo" is better than ChildIpv6AddressList as it includes the RLOC integer.

Reply:
I think the previous definition of the `ChildIpv6AddressList ` is incorrect.
it should be defined as follow:
```
std::vector<ChildIpv6AddressInfo> mChildIpv6AddrsInfoList;
```
`ChildIpv6AddressInfo` represents the address information recorded by a child of the device, including its RLOC (Routing Locator) and addresses.  std::vector<ChildIpv6AddressInfo> records the address information of all the children of the device.